### PR TITLE
SymPy unit tests

### PR DESCRIPTION
**Pull request by hsiang-wu**
Unit tests of SymPy showing its basic usages.

New file `simple_test.py` contains simple tests of SymPy functions.
New file `simplification_test` showcases simplification-related functions of SymPy.

### CODE REVIEW

File: neuralpp/test/quick_tests/symoblic/symbols_test.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+https://docs.sympy.org/latest/tutorial/gotchas.html
+"""
+from sympy import *
+
+
+def test_symbols_basic() -> None:
+    """SymPy supports simple expression conversions. e.g., `x + 2 * y` is the same as `y + x + y`"""
+    x, y = symbols('x y')
+    expr = x + 2 * y
+    assert expr == x + 2 * y
```
Comment (rodrigodesalvobraz) [inline neuralpp/test/quick_tests/symoblic/symbols_test.py]: Interesting! Does that mean the operator == does normalization work every time, or are expressions immediately converted to a normal form as soon as they are created so that equivalent functions are represented identically internally? Does the manual say anything about that? I guess one way to find out is to create two equivalent but distinct expressions and print them to see if their string representation is the same (which would mean they are normalized and the initial form is lost).

File: neuralpp/test/quick_tests/symoblic/simplication_test.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+Tests of SymPy's simplication functions.
+https://docs.sympy.org/latest/tutorial/simplification.html
+"""
+from sympy import *
+
+
+def test_simplify() -> None:
+    """
+    One can just use sympy.simplify().
+    It's push-button but has 2 caveats:
```
Comment (rodrigodesalvobraz) [inline neuralpp/test/quick_tests/symoblic/simplication_test.py]: What does "push-down" mean?

File: neuralpp/test/quick_tests/symoblic/simplication_test.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    1. "simplest" form is not well-defined. (showcased by last set of asserts)
+    2. it can be unnecessarily slow due to generality.
+    """
+    x = symbols('x')
+    assert simplify(sin(x)**2 + cos(x)**2) == 1
+    assert simplify((x**3 + x**2 - x - 1)/(x**2 + 2*x + 1)) == x - 1
+    assert simplify(gamma(x) / gamma(x - 2)) == (x - 2) * (x - 1)
+    # we may want to simplify (x**2 + 2*x + 1) into (x+1)**2 but that's not "simplest" recognized by simplify()
```
Comment (rodrigodesalvobraz) [inline neuralpp/test/quick_tests/symoblic/simplication_test.py:19]: It's very nice that you are using the tests to illustrate the finer points like that!

File: neuralpp/test/quick_tests/symoblic/simplication_test.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+def test_expand() -> None:
+    """ The function expand() expands. """
+    x = symbols('x')
+    assert expand((x + 2)**2) == x**2 + 4*x + 4
+    assert expand((x + 2)*(x - 2)) == x**2 - 4
+
+
+def test_factor() -> None:
+    """ The function factor() turns a polynomial into an irreducible product of factors. """
```
Comment (rodrigodesalvobraz) [inline neuralpp/test/quick_tests/symoblic/simplication_test.py]: Does it work for arbitrary degrees? I don't know much about this but my understanding is that this is hard for higher degrees.

File: neuralpp/test/quick_tests/symoblic/simple_test.py [hunk 5]
```diff
@@ -54,18 +57,37 @@ def test_change_after_create() -> None:
...
+
+
+def test_eval():
+    """ Test numerical evaluation, including a quantifier example. """
+    x, i = symbols("x i")
+    assert sqrt(8) != math.sqrt(8)
+    assert sqrt(8).evalf() == math.sqrt(8)
+    assert cos(2*x).evalf(subs={x: 2.4}) == math.cos(2*2.4)
+
+    expr = Sum(Indexed('x', i), (i, 0, 3))  # expr is a quantified expression
```
Comment (rodrigodesalvobraz) [inline neuralpp/test/quick_tests/symoblic/simple_test.py:94]: Nice that you tested a quantifier! But I don't quite get the meaning here: is i the index? How is x being used exactly? What is the meaning of (i, 0, 3)?

File: neuralpp/test/quick_tests/symoblic/simplication_test.py [hunk 6]
```diff
@@ -22,33 +21,53 @@ def test_simplify() -> None:
     assert factor(x**2 + 2*x + 1) == (x + 1)**2
 
 
-def test_expand() -> None:
-    """ The function expand() expands. """
-    x = symbols('x')
+def test_polynomial_rational() -> None:
```
Comment (rodrigodesalvobraz) [inline neuralpp/test/quick_tests/symoblic/simplication_test.py:25]: The original name test_expand seemed clearer to me, I guess because "expand" is an action and we test actions. What does it mean to test a polynomial_rational?

File: neuralpp/test/quick_tests/symoblic/simple_test.py [hunk 7]
```diff
@@ -83,11 +83,16 @@ def test_substitution():
...
     assert sqrt(8) != math.sqrt(8)
     assert sqrt(8).evalf() == math.sqrt(8)
     assert cos(2*x).evalf(subs={x: 2.4}) == math.cos(2*2.4)
 
     expr = Sum(Indexed('x', i), (i, 0, 3))  # expr is a quantified expression
-    expr_as_func = lambdify(x, expr)
+    expr_as_func = lambdify(x, expr)  # lambdify() turns the expression into a python function
     assert expr_as_func([1, 2, 3, 4, 5]) == 10
+
+    # doit() can also do computation/simplification, note the interval can be infinite
```
Comment (rodrigodesalvobraz) [inline neuralpp/test/quick_tests/symoblic/simple_test.py:98]: What is the different between doit and simplification? When should we use one or the other?

File: neuralpp/test/quick_tests/symoblic/simple_test.py [hunk 8]
```diff
@@ -73,15 +76,15 @@ def test_gotchas_xor_and_divide() -> None:
     assert Rational(1, 3) == Integer(1)/Integer(3)
 
 
-def test_substitution():
+def test_substitution() -> None:
```
Comment (rodrigodesalvobraz) [inline neuralpp/test/quick_tests/symoblic/simple_test.py:79]: Why did you add "-> None"? Was that required?

File: neuralpp/test/quick_tests/symoblic/symbols_test.py (hunk 1 above)
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/symbols_test.py]: > Does that mean the operator == does normalization work every time, or are expressions immediately converted to a normal form as soon as they are created so that equivalent functions are represented identically internally?

I remember I read from the doc somewhere that it is the latter, where they gave the example of `x + 1` and `1 + x` are printed in the same way. And for `==`, it is said that "in SymPy, == represents exact structural equality testing."

File: neuralpp/test/quick_tests/symoblic/simplication_test.py (hunk 4 above)
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/simplication_test.py]: Neither do I. From the tutorial:

> factor() uses a complete multivariate factorization algorithm over the rational numbers, which means that each of the factors returned by factor() is guaranteed to be irreducible.

So my guess is it works for arbitrary degrees though it might be slower for higher degrees.

File: neuralpp/test/quick_tests/symoblic/simple_test.py (hunk 5 above)
Comment (rodrigodesalvobraz) [reply neuralpp/test/quick_tests/symoblic/simple_test.py:94]: Yes, I think that deserves some more detail (enough to understand how to use it ourselves).

File: neuralpp/test/quick_tests/symoblic/simplication_test.py (hunk 6 above)
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/simplication_test.py:25]: A better name may be `test_polynomial_or_rational_function_simplification`. The library has a dedicated set of simplification functions for each kind of expressions, e.g., `expand()` and `factor()` for polynomial/rational, `trigsimp()` for trigonometric.

File: neuralpp/test/quick_tests/symoblic/simple_test.py (hunk 7 above)
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/simple_test.py:98]: I should have avoid using the word "simplification". A better word would be "evaluation". Since the expression is symbolic, `Sum` is just `Sum`, they don't realy "do" the sum. Calling `doit()` would actually do the sum.

So calling `doit()` makes sense on the expressions that you can "do", such as `Sum` and `Integral`. To use another example from the doc, `2*Integral(x, x)` is just `2*Integral(x, x)`, but `(2*Integral(x, x)).doit()` would do the integral and result in `x**2`.

File: neuralpp/test/quick_tests/symoblic/simple_test.py (hunk 8 above)
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/simple_test.py:79]: No. Just want to write it in a more typed way..

File: neuralpp/test/quick_tests/symoblic/simple_test.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+def test_eval() -> None:
+    """ Test numerical evaluation, including a quantifier example. """
+    x, i, k = symbols("x i k")
+    assert sqrt(8) != math.sqrt(8)
+    assert sqrt(8).evalf() == math.sqrt(8)
+    assert cos(2*x).evalf(subs={x: 2.4}) == math.cos(2*2.4)
+
+    expr = Sum(Indexed('x', i), (i, 0, 3))  # expr is a quantified expression
+    expr_as_func = lambdify(x, expr)  # lambdify() turns the expression into a python function
+    assert expr_as_func([1, 2, 3, 4, 5]) == 10
```
Comment (CactusWin) [inline neuralpp/test/quick_tests/symoblic/simple_test.py:96]: Is there a reason that you included `5` in the list,  since it won't be reached in the expression? Was it to show that it'll ignore extra arguments?

**Review by CactusWin — COMMENTED** [review]
Very nice! I really liked the comments you made! It made the test much easier to read and understand. :) I'll approve once the changes that Rodrigo suggested are made/ resolved.

File: neuralpp/test/quick_tests/symoblic/simple_test.py (hunk 8 above)
Comment (rodrigodesalvobraz) [reply neuralpp/test/quick_tests/symoblic/simple_test.py:79]: Yes, I think it may be distracting since the rest of the code is not typed.

File: neuralpp/test/quick_tests/symoblic/simplication_test.py (hunk 2 above)
Comment (rodrigodesalvobraz) [reply neuralpp/test/quick_tests/symoblic/simplication_test.py]: Ah, now I understand. I think many people may get confused like I did, so can you please add a short explanation ("automatic and general", something like that)?

File: neuralpp/test/quick_tests/symoblic/simplication_test.py (hunk 6 above)
Comment (rodrigodesalvobraz) [reply neuralpp/test/quick_tests/symoblic/simplication_test.py:25]: Can you please update the name then? Thanks.

File: neuralpp/test/quick_tests/symoblic/simple_test.py (hunk 7 above)
Comment (rodrigodesalvobraz) [reply neuralpp/test/quick_tests/symoblic/simple_test.py:98]: Can you please add some of that explanation to the comments? Thanks.

**Review by horizon-blue — COMMENTED** [review]
:D I really like the detailed comments. The two test files looks great to me overall minus Rodrigo's and Winnie's suggestions. Once we address those, the PR should be good to go.

File: neuralpp/test/quick_tests/symoblic/simple_test.py (hunk 9 above)
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/simple_test.py:96]: Yes. To show `i` indexes from 0 to 3 and not beyond.

File: neuralpp/test/quick_tests/symoblic/simplication_test.py (hunk 2 above)
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/simplication_test.py]: I replaced "push-button" with "general", does that make sense?

File: neuralpp/test/quick_tests/symoblic/simple_test.py (hunk 5 above)
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/simple_test.py:101]: Created a new test function for `Sum()`

File: neuralpp/test/quick_tests/symoblic/simplication_test.py (hunk 6 above)
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/simplication_test.py]: Changed the function names in this file.

File: neuralpp/test/quick_tests/symoblic/simple_test.py (hunk 7 above)
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/simple_test.py]: Refactored test for `Sum()` and moved test of `doit()` there;
switched to the term "symbolic evaluation" in the comment.

File: neuralpp/test/quick_tests/symoblic/simple_test.py (hunk 8 above)
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/simple_test.py]: Got rid of `-> None` in test files.

**Comment by CactusWin** [comment]
This looks good to me! I'm not too sure how to approve the PR, since the review button seemed to have disappeared for me :)

**Review by horizon-blue — APPROVED** [review]
@CactusWin I think you can either re-request review for your account from the right side panel, or visit the "Files changed" tab to directly create a new review :)

File: neuralpp/test/quick_tests/symoblic/simple_test.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+def test_sum():
+    """ A more detailed test of the Sum quantifier.
+    In quote of Sum's doc, the first argument is "the general form of terms in the series" and
+    the second argument is "(dummy_variable, start, end), with dummy_variable taking all
+    integer values from start through end. In accordance with long-standing mathematical convention,
+    the end term is included in the summation."
+    """
+    x, i, k = symbols("x i k")
+    # In the following example, `i` is not used in the series but just to indicate the index of 'x'.
+    expr = Sum(Indexed('x', i), (i, 0, 100))
```
Comment (rodrigodesalvobraz) [inline neuralpp/test/quick_tests/symoblic/simple_test.py:115]: I think this is a bit too confusing for a first example. I thought that somehow `Indexed` referred to the summation's index, not that `x` was a list being indexed. Also, I got confused by your comment that `i` is not used in the series, because it is (to index `x`). It would be good to start with a plainer example like `Sum(2*i, (i, 0, 100))` and then move on to the use of `Indexed`.

**Review by rodrigodesalvobraz — CHANGES_REQUESTED** [review]
Looking good! Please address a small change I suggested for the Sum function and then we should merge. Thanks!

File: neuralpp/test/quick_tests/symoblic/simple_test.py (hunk 10 above)
Comment (hsiang-wu) [reply neuralpp/test/quick_tests/symoblic/simple_test.py:120]: Got it. I'v reordered tests and modified comments.

**Review by rodrigodesalvobraz — APPROVED** [review]
